### ISSUE TITLE
from sklearn.metrics import silhouette_score, using metric = 'precomputed'

### ISSUE THREAD

**Issue by jayypaul** [scverse/scanpy#222]
After pp.neighbors and tl.louvain, I've been calculating the silhouette index of the clustering arrangements to get an idea of how well the data is clustered: sil_avg = silhouette_score(adata.X, adata.obs['louvain'], metric = 'euclidean'). I'm using a for loop to go through many combinations of paramaters, (e.g. n_neighbors for pp.neighbors and resolution for tl.louvain), but it seems that the silhouette indices being generated aren't what I would expect. 

However, I've noticed that there is an option, metric ='precomputed', which can be used if a distance matrix has already been calculated, which is done so after using pp.neighbors : adata.uns['neighbors']['distances']. 

Any idea how to use silhouette_score with adata.uns['neighbors']['distances']? 
I've tried a couple different ways, e.g. : 
silhouette_score(adata.X, adata.uns['neighbors']['distances'], metric='precomputed').

http://scikit-learn.org/stable/modules/generated/sklearn.metrics.silhouette_score.html

**Comment by falexwolf** [comment]
The biggest problem here is that you take `adata.X` in the `silhouette_score` function to compute distances. However, the basis for the clustering typically is the PCA representation of data in  `adata.obsm['X_pca']`. If you adopt this, you should get meaningful results.

Alternatively, you should be able to pass `adata.uns['neighbors']['distances']` as the first argument of `silhouette_score` if you pass `metric='precomputed'` at the same time... This would also solve the problem I mentioned above. Maybe it doesn't accept sparse matrices and you need to run `.toarray()` before. You don't need `adata.X` in this case, either.

**Comment by jayypaul** [comment]
`adata.obsm['X_pca']` seems to have predicted better clustering. Thank you. I'm going to try that with the rest of my data and see what clustering it suggests. 

As for `adata.uns['neighbors']['distances']`, shortly after my post I read that the `silhouette_score` function wasn't designed to handle spares matrices, so you're right with that. 
(source: https://books.google.com/books?id=skvZDQAAQBAJ&pg=PA786&lpg=PA786&dq=silhouette_score+precomputed+python3&source=bl&ots=YRC9VPTPPW&sig=7KPSQDWtZG6537-f_vZGwMpMvCc&hl=en&sa=X&ved=2ahUKEwjO3ruPqsXcAhWEg-AKHXD5BUAQ6AEwCXoECAMQAQ#v=onepage&q=silhouette_score%20precomputed%20python3&f=false). 
It suggested using todense() if the matrix is small, but to avoid this function all together if the matrix is large. When I trade toarray() for todense() it seems to produce similar results. Since single-cell datasets are likely always to be too big, it suggested using V-Measure or Adjusted Mutual Information as a way to evaluate sparse matrices instead. Just thought I'd update with my findings.

Again, the `adata.obsm['X_pca']` suggestion seems to predict better clustering arrangements. Big ups for that! 

Best
